File: client/src/client/api/Subscription.ts
import {OfflineNotifier} from "./OfflineNotifier";

export interface Subscription {
    destination: string;

    cancel(): void;

    send(destination: string,
         content: string,
         headers: { [key: string]: string },
         contentType: string
    ): void


    isOffline: boolean;

    onOfflineChanged(notifier: (newValue: boolean) => void): Subscription;

    addOfflineNotifier(notifier: OfflineNotifier): Subscription;

    removeOfflineNotifier(notifier: OfflineNotifier): Subscription;
}

File: server/src/main/kotlin/server/Subscription.kt
package server

interface Subscription {
    val destination: String

    fun cancel()

    fun send(
        destination: String,
        content: String,
        headers: Map<String, String> = mapOf(),
        contentType: String = "text/plain"
    )

    val isOffline: Boolean
    fun onOfflineChanged(notifier: (newValue: Boolean) -> Unit): Subscription;
    fun addOfflineNotifier(notifier: OfflineNotifier): Subscription;
    fun removeOfflineNotifier(notifier: OfflineNotifier): Subscription;
}